### PR TITLE
Webpack bug fixes

### PR DESCRIPTION
**Pull request by SaemJeon**
Create tests for webpack plugins

### CODE REVIEW

File: tools/webpack-core/src/test/RequireMagicCommentsPlugin.test.ts [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+  });
+
+  /* Test A */
+  it("A: sucess with external comments with custom handler", async () => {
+    testConfig =
+      getRequireExternalConfig(
+        "assets/require-magic-comments-plugin-test/a/requireTestA.js",
+        [new RequireMagicCommentsPlugin([
+          {
+            test: /webpack: *external/i,
```
[the remark this answers is not in the document]
Comment (wgoehrig) [reply tools/webpack-core/src/test/RequireMagicCommentsPlugin.test.ts:47]: This also goes for tests C & D

File: tools/webpack-core/src/test/RequireMagicCommentsPlugin.test.ts [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    expect(logging[0].type).to.be.equal("log");
+    expect(logging[0].message).to.include(`Handler for /*webpack: external*/ - transformed "foo" => "BeWebpack-EXTERNAL:foo" at lib\\test\\assets\\require-magic-comments-plugin-test\\b\\requireTestB.js`);
+
+    const testModulePath = path.join(__dirname, "dist/requireTestB.js");
+    const testModuleContent = fs.readFileSync(testModulePath, "utf8");
+    expect(testModuleContent).to.be.equal(`(window["webpackJsonp"] = window["webpackJsonp"] || []).push([[0],[\n/* 0 */\n/***/ (function(module, exports, __webpack_require__) {\n\nmodule.exports = __webpack_require__(1);\n\n\n/***/ }),\n/* 1 */\n/***/ (function(module, exports, __webpack_require__) {\n\n__webpack_require__(/*webpack: external*/2);\n\n/***/ }),\n/* 2 */\n/***/ (function(module, exports) {\n\nmodule.exports = require("foo");\n\n/***/ })\n],[[0,1]]]);`);
+  });
+
+  /* Test C */
+  it("C: sucess with external comments with with resolve not converted", async () => {
```
Comment (wgoehrig) [inline tools/webpack-core/src/test/RequireMagicCommentsPlugin.test.ts:99]: nit: typo (with with)

same for tests D-F

File: tools/webpack-core/src/test/CopyExternalsPlugin.test.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    const result = await new Promise<any>((resolve, reject) => {
+      testCompiler.run((err: any, stats: any) => (err) ? reject(err) : resolve(stats));
+    });
+
+    const logging = result.toJson({ logging: true }).logging.CopyExternalsPlugin.entries;
+    expect(logging.length, "Length of entries should be 2").to.be.equal(2);
+    expect(logging[0].type).to.be.equal("warn");
+    expect(logging[0].message).to.be.equal(`Can't copy package "eslint" - it is not a direct dependency.`);
+    expect(logging[1].type).to.be.equal("log");
+    expect(logging[1].message).to.be.include(`"eslint" included at lib\\test\\assets\\copy-externals-plugin-test\\copyEslintTest.js`);
```
[the remark this answers is not in the document]
Comment (wgoehrig) [reply tools/webpack-core/src/test/CopyExternalsPlugin.test.ts:59]: This probably applies to all of these logging assertions...

**Review by wgoehrig — CHANGES_REQUESTED** [review]

File: tools/webpack-core/src/test/OptionalDependenciesPlugin.test.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+      "lib/test/assets/optional-dependencies-plugin-test/b/node_modules/optional-b/index.js": `try {\n  const b = require("./b");\n} catch {\n\n}`,
+      "lib/test/assets/optional-dependencies-plugin-test/b/node_modules/optional-b/b.js": `console.log("Test B");`,
+      "lib/test/assets/optional-dependencies-plugin-test/c/optionalTestC.js": `const optionalC = require("optional-c");`,
+      "lib/test/assets/optional-dependencies-plugin-test/c/node_modules/optional-c/index.js": `const pathC = "./c";\nconst c = require(pathC);`,
+      "lib/test/assets/optional-dependencies-plugin-test/c/node_modules/optional-c/c.js": `console.log("Test C");`,
+      "lib/test/assets/optional-dependencies-plugin-test/d/optionalTestD.js": `const optionalD = require("optional-d");`,
+      "lib/test/assets/optional-dependencies-plugin-test/d/node_modules/optional-d/index.js": `const optional = true;\nlet req = (optional) ? require : console.log;\nconst d = req("./d");`,
+      "lib/test/assets/optional-dependencies-plugin-test/d/node_modules/optional-d/d.js": `console.log("Test D");`,
+    });
+    /* eslint-enalbe @typescript-eslint/naming-convention */
```
[the remark this answers is not in the document]
Comment (saemjeon) [reply tools/webpack-core/src/test/OptionalDependenciesPlugin.test.ts:34]: Naming convention for node_modules I believe. It was asking for camelCase.

File: tools/webpack-core/src/test/CopyExternalsPlugin.test.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  });
+
+  it("should copy @bentley/imodeljs-native", async () => {
+    vol.fromJSON({
+      "lib/test/assets/copy-externals-plugin-test/test.js": `import * as imodeljsnative from "@bentley/imodeljs-native";`,
+    });
+    testConfig = getTestConfig("assets/copy-externals-plugin-test/test.js", [new CopyExternalsPlugin()], ["@bentley/imodeljs-native"]);
+
+    await runWebpack(testConfig, ufs);
+    expect(fs.existsSync(path.join(__dirname, "dist/node_modules/@bentley/imodeljs-native"))).to.be.true;
```
[the remark this answers is not in the document]
Comment (SaemJeon) [reply tools/webpack-core/src/test/CopyExternalsPlugin.test.ts:70]: I believe it is `@bentley/imodeljs-native` is found from the regular node_modules directory.

[the remark this answers is not in the document]
Comment (wgoehrig) [reply tools/webpack-core/src/test/CopyExternalsPlugin.test.ts:61]: OK, let's create a "fake" `@bentley/imodeljs-native` in `lib/test/assets/copy-externals-plugin-test/node_modules/` then...

I don't like the tests resolving anything in the root node_modules dir

[the remark this answers is not in the document]
Comment (wgoehrig) [reply tools/webpack-core/src/test/CopyExternalsPlugin.test.ts]: @SaemJeon I thought we agreed to just remove this test?

**Review by wgoehrig — APPROVED** [review]

**Pull request merged**